Add review request notification after 100 clicks

danielbitzer opened the pull request:
### Changes proposed in this Pull Request:

This PR adds the first review request notification from #860.

The note should display after 100 free listing clicks but it rounds down to the nearest hundred so if a site has 230 clicks it will display "200" in the note. The **Leave a review** link randomizes between WC.com and WP.org as per discussion on the issue. 

### Questions: 

- Since notes currently register on every `admin_init` action, we need to either use a transient to cache the clicks count or implement a cron task that register's notes daily. It looks like we will be adding more notes and some of our notes run order queries so I think it makes sense to implement a plugin cron heartbeat that runs daily. 
- Unsure whether `where_date_between` should accept `DateTime` objects or strings. I've gone with strings for now.

Helps #860.

### Screenshots:

<img width="539" alt="Screen Shot 2021-09-17 at 1 28 25 pm" src="https://user-images.githubusercontent.com/1892137/133721946-8a7bd2d9-2bb4-49ab-ad96-50217d564ee0.png">

### Detailed test instructions:

1. If you have more than 100 clicks the notice should appear right away
2. Otherwise you mock date from the connect server (personally I haven't done this yet but y'all are probably experts at this) p1631770014256800-slack-CK365S85V 


### Changelog entry

> Add - Review request inbox notification after 100 free listing clicks

danielbitzer committed:
Add new MerchantMetrics service and initial clicks count

danielbitzer committed:
Add number_format_i18n() proxy method

danielbitzer committed:
Add ReviewAfterClicks note

danielbitzer committed:
Remove unused imports

danielbitzer committed:
Bail if there are no results

danielbitzer committed:
Remember to check if the note can be added

danielbitzer committed:
tiny phpcs fix

danielbitzer committed:
Tiny cosmetic tweaks

danielbitzer committed:
Merge branch 'develop' into feature/860-review-notes

danielbitzer committed:
Remove unused trait

danielbitzer committed:
Tweak variable names

danielbitzer committed:
Add Query::where_date_between shortcut method

danielbitzer committed:
Build query using methods rather than an array

danielbitzer committed:
Add wp_timezone() to WP proxy

danielbitzer committed:
Convert date args objects to strings

danielbitzer committed:
Update where_date_between docblock

danielbitzer committed:
Retreive timezone from WP proxy

danielbitzer committed:
phpcs fixes

danielbitzer committed:
Simplfy where_date_between() return

danielbitzer committed:
Add unit tests for get free listing metric

layoutd commented on src/API/Google/MerchantMetrics.php:
```diff
@@ -0,0 +1,83 @@
+	 *
+	 * @return string
+	 */
+	protected function get_today(): string {
```
I wonder if this would be useful in other places, for example:
https://github.com/woocommerce/google-listings-and-ads/blob/bd62aaed77f591e3d7f69d6b9b01625adb5c0eca/src/API/Google/Proxy.php#L649-L655
(Maybe with an indication of the format of the return string)

danbitzer replied:
In that example a different timezone is used. I think there is definitely room for some datetime helper code in this project. I played around with some ideas while working on this PR but didn't arrive on anything.

layoutd replied:
> In that example a different timezone is used. 

❓  Sorry, end of the week, couldn't quite parse that. 

But I'd agree that some datetime helper code would be good, maybe add it as an issue with some notes about what you were thinking and a reference to this method for refactoring?

layoutd commented on src/API/Site/Controllers/BaseReportsController.php:
```diff
@@ -155,7 +155,7 @@ protected function get_defaults( array $params ): array {
 	protected function normalize_timezones( &$query_args ) {
 		/** @var WP $wp */
 		$wp       = $this->container->get( WP::class );
-		$local_tz = new DateTimeZone( $wp->wp_timezone_string() );
```
Can remove `import DateTimeZone` with this change.

layoutd commented on src/Internal/DependencyManagement/CoreServiceProvider.php:
```diff
@@ -232,6 +235,7 @@ function ( ...$arguments ) {
 		// Inbox Notes
 		$this->conditionally_share_with_tags( ContactInformationNote::class );
 		$this->conditionally_share_with_tags( CompleteSetupNote::class );
+		$this->conditionally_share_with_tags( ReviewAfterClicksNote::class, MerchantMetrics::class, WP::class );
```
As @nima-karimi [astutely pointed out](https://github.com/woocommerce/google-listings-and-ads/pull/1016#discussion_r716578920) in another class, here the `Notes` classes and their subclasses don't actually implement `Conditional`, so there's no need to use `conditionally_share_with_tags`.

danbitzer replied:
Good point!

layoutd commented on src/Proxies/WP.php:
```diff
@@ -138,4 +139,31 @@ public function is_wp_error( $thing ): bool {
+	 *
+	 * @since x.x.x
+	 */
+	public function wp_timezone() {
```
Missing return type declaration.

danielbitzer replied:
It's there, just 2 lines above.

layoutd reviewed (commented):
This looks good and tests well! I left a few comments but nothing blocking.

> Since notes currently register on every `admin_init` action, we need to either use a transient to cache the clicks count or implement a cron task that registers notes daily. It looks like we will be adding more notes and some of our notes run order queries so I think it makes sense to implement a plugin cron heartbeat that runs daily.

Agree that a plugin cron heartbeat looks like the right choice (or at the very least a transient) – sending a request to the API (and thus through the WCS) for every `admin_init` seems like a recipe for problems. I was also able to queue up [multiple instances of the same note] by hitting `CMD+R` multiple times when the API returned a value over 100:
![image](https://user-images.githubusercontent.com/228780/135109867-e46ee9a7-5110-4c2e-ac60-5f3eeffc2e87.png)

danielbitzer committed:
Add function return type

danielbitzer committed:
Remove unused import

danielbitzer committed:
Notes aren't `Conditional`

danielbitzer committed:
Merge branch 'develop' into feature/860-review-notes

danielbitzer committed:
Merge branch 'develop' into feature/860-review-notes

- Resolve changes to Note structure
- Add exception handling and caching for clicks merchant metric

danielbitzer committed:
phpcs

danielbitzer committed:
Remove duplicate conditional

danielbitzer commented:
Thanks for the review @layoutd! I've made some tweaks:
- updated based on the changes in #1046 
- I noted `MerchantFreeListingReportQuery` could throw exceptions so I've added some exception handling but I'm not 100% sure if I've understood how exceptions should be handled in this project. I opted to catch late and use the `do_action` hook for logging.
- Added some local caching of the clicks count in the Note class, because I had retrieve the count twice in two different locations in the class

danielbitzer committed:
Merge branch 'develop' into feature/860-review-notes

danielbitzer committed:
Use tomorrow's date to include metrics from current day

danielbitzer committed:
Minor cleanup

layoutd commented on src/Notes/ReviewAfterClicks.php:
```diff
@@ -0,0 +1,140 @@
+			return false;
+		}
+
+		return true;
```
```suggestion
		return $clicks_count > 100;
```

layoutd reviewed (changes requested):
The Note works well, but running the tests on this are giving me a seg fault and it looks like it's also happening in the GH Action tests: https://github.com/woocommerce/google-listings-and-ads/runs/4030983635?check_suite_focus=true

As an aside the NoteInitializer `deactivate` method breaks plugin deactivation if the AS task is manually cancelled. (Cancel `wc_gla_cron_daily_notes` in Pending Schedule Actions wp-admin/tools.php?page=action-scheduler&status=pending`, then deactivate the plugin).  Adding a `has_scheduled_action` check before trying to cancel the job seems to resolve the issue.

danielbitzer committed:
Merge branch 'develop' into feature/860-review-notes

danielbitzer committed:
Don't mock WP class to fix tests

danielbitzer committed:
Update expected query date range

danielbitzer committed:
Catch and log exceptions from `ActionSchedulerInterface::cancel`

danielbitzer committed:
No need for WP instance to be a prop

danielbitzer committed:
Shorten conditional

Co-authored-by: Justin Palmer <justin.palmer@automattic.com>

danielbitzer commented:
> a seg fault and it looks like it's also happening in the GH Action tests

Haven't seen that issue before! Looks like it was due mocking `WP` class rather than just passing a real one.

> As an aside the NoteInitializer deactivate method breaks plugin deactivation if the AS task is manually cancelled. 

Good catch! I didn't realise `\Automattic\WooCommerce\GoogleListingsAndAds\ActionScheduler\ActionSchedulerInterface::cancel` throws exceptions. I opted to catch and log the exception rather than add an checking the action existed since the action definitely should exist.

danielbitzer committed:
Fix typo

danielbitzer committed:
Expand free listing clicks metric function to include impressions

Also adds transient caching method for free listing metrics.

danielbitzer committed:
Add test_get_free_listing_metrics_with_no_merchant_id

layoutd commented on src/API/Google/MerchantMetrics.php:
```diff
@@ -0,0 +1,126 @@
+		$report_row = $response->getResults()[0];
+
+		return [
+			'clicks' => (int) $report_row->getMetrics()->getClicks(),
+			'impressions' => (int) $report_row->getMetrics()->getImpressions(),
```
Assignment alignment
```suggestion
			'clicks'      => (int) $report_row->getMetrics()->getClicks(),
			'impressions' => (int) $report_row->getMetrics()->getImpressions(),
```

layoutd commented on src/Notes/ReviewAfterClicks.php:
```diff
@@ -0,0 +1,130 @@
+	/**
+	 * Checks if a note can and should be added.
+	 *
+	 * - checks that the plugin is setup
```
This method itself doesn't test that the plugin is set up, but it is part of its process 🤷🏻

layoutd reviewed (changes requested):
> Haven't seen that issue before!

Yeah, I also haven't seen that in a long time (if at all 🤔 ). Thanks for sorting it out. And everything else looks good, using `try-catch` is much better for the deactivation exception.

The only (final!) ~squibble~ squabble I have is that `NoteInitializer::install()` is invoked on plugin install or update, but the `wc_gla_cron_daily_notes` hook is removed on deactivation. So during testing I deactivated the plugin, then reactivated, but task wasn't re-scheduled.

danielbitzer committed:
Ensure note cron job is added if the plugin is reactivated

danielbitzer committed:
phpcs

danielbitzer committed:
Update comment

danielbitzer commented:
Thanks @layoutd! I've updated as per your feedback. Since the changes are pretty minor I'll go ahead and merge.

> The only (final!) squibble I have is that `NoteInitializer::install()` is invoked on plugin install or update, but the `wc_gla_cron_daily_notes` hook is removed on deactivation. So during testing I deactivated the plugin, then reactivated, but task wasn't re-scheduled.

Another great catch!
